helpers/subscriptionControllers.js

```diff
@@ -0,0 +1,140 @@
+		this.updateSubscribers();
+		if (this._callbacks.onSubscribe) this._callbacks.onSubscribe(target);
+	}
+
+	unsubscribe(target) {
+		this.subscribers.delete(target);
+		if (this._callbacks.onUnsubscribe) this._callbacks.onUnsubscribe(target);
+	}
+
+	updateSubscribers() {
+		if (!this.subscribers || this.subscribers.size === 0) return;
+		if (!this._callbacks.updateSubscribers) return;
+
+		// debounce the updates
+		if (this._updateSubscribersRequested) return;
+
+		this._updateSubscribersRequested = true;
+		setTimeout(() => {
+			this._callbacks.updateSubscribers(this.subscribers);
+			this._updateSubscribersRequested = false;
+		}, 0);
+	}
+
+	_handleSubscribe(e) {
+		e.stopPropagation();
+		e.detail.provider = this._host;
+		const target = e.composedPath()[0];
+		this.subscribe(target);
+	}
+}
+
+export class SubscriberController {
+
+	constructor(host, callbacks, options) {
+		this._host = host;
+		this._callbacks = callbacks;
+		this._eventName = options.eventName;
+		this._forPropertyName = options.forProperty;
+		this._controllerId = options.controllerId;
+		this._providers = new Map();
+	}
+
+	// This method is basically offered for convenience
+	// Components using the event method or a forProperty that takes a string will know they only have one provider
+	get provider() {
+		return this._providers.values().next().value;
+	}
+
+	get providers() {
+		return Array.from(this._providers.values());
+	}
+
+	hostConnected() {
+		if (!this._eventName || this._host[this._forPropertyName]) return;
+		// delay subscription otherwise import/upgrade order can cause selection mixin to miss event
+		requestAnimationFrame(() => {
+			const evt = new CustomEvent(this._eventName, {
+				bubbles: true,
+				composed: true,
+				detail: {}
+			});
+			this._host.dispatchEvent(evt);
+			// When using the event method, the provider is not guarenteed to have an id so we use the provider as the key here
+			this._providers.set(evt.detail.provider, evt.detail.provider);
+			if (this._callbacks.onSubscribe) this._callbacks.onSubscribe(evt.detail.provider);
+		});
+	}
+
+	hostDisconnected() {
+		if (this._forObserver) this._forObserver.disconnect();
+		this._providers.forEach(provider => {
+			provider.getController(this._controllerId).unsubscribe(this._host);
+		});
+	}
+
+	hostUpdated(changedProperties) {
+		if (!changedProperties.has(this._forPropertyName)) return;
+
+		if (this._forObserver) this._forObserver.disconnect();
+		this._providers.forEach(provider => {
+			provider.getController(this._controllerId).unsubscribe(this._host);
+		});
+		this._providers = new Map();
+
+		this._updateProviders();
+
+		this._forObserver = new MutationObserver(() => {
+			this._updateProviders();
+		});
+
+		this._forObserver.observe(this._host.getRootNode(), {
+			childList: true,
+			subtree: true
+		});
+	}
+
+	_updateProviders() {
+		let providerIds = this._host[this._forPropertyName];
+		if (typeof(providerIds) === 'string') providerIds = [providerIds];
```

Comment:
Shouldn't it always be a string? If we want to support multiple I'd expect it to still be a string but have the IDs be space-separated.

Reply:
Interesting, I automatically assumed it'd be an array of strings.  But string with spaced ids works too

Reply:
I'm curious what Lit's default attribute processor does for arrays. Like would it turn `<d2l-foo ids="one two three">` into an array if that property had an `Array` type? I'm good with the property being an array as long as that's how the parsing works (which matches what HTML does for things like `aria-labelledby`).

Reply:
So looks like no - if I set `type` to `Array` and pass in a string, things blow up.  It's trying to run `JSON.parse(<string_I_pass_in>)`.  It looks like I should be able to write a custom converter to take in the string and convert it to an array.  Or we just expect a spaced string and handle it, or just expect an array and handle it.

Reply:
Yeah I was just thinking about those custom parsers -- if they're easy to write that could be nice.